src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -2532,4 +2535,154 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
+        r.clear()
+        rb.push(r)
+      }
+
+      val deque = new java.util.ArrayDeque[(Locus, RegionValue)]()
+
+      it.flatMap { f =>
+        // these rows are loaded from previous partitions
+        val inAdjustment = it.hasNext
+        val rvIterator = f(newContext)
+
+        val region = context.region
+        val rv2 = RegionValue()
+        val rvb = new RegionValueBuilder()
+        rvIterator.flatMap { rv =>
+          val locus = UnsafeRow.readLocus(rv.region, localRVRowType.loadField(rv, locusIndex), rg)
+
+          val result = if (inAdjustment) {
+            None
+          } else {
+            def discard(x: (Locus, RegionValue)): Boolean = x != null && (x._1.position < locus.position - basePairs
+              || x._1.contig != locus.contig)
+
+            while (discard(deque.peekLast()))
+              recycleRegion(deque.removeLast()._2.region)
+
+            val rvs = deque.iterator().asScala.map(_._2).toArray
+
+            rvb.set(region)
+            rvb.clear()
+            region.clear()
+            rvb.start(newType.rvRowType)
+            rvb.startStruct()
+            rvb.addFields(localRVRowType, rv, nonEntryIndices)
+
+            // prev_rows
+            rvb.startArray(rvs.length)
+            var j = 0
+            while (j < rvs.length) {
+              val rvj = rvs(j)
+              rvb.startStruct()
+              rvb.addFields(localRVRowType, rvj, nonEntryIndices)
+              rvb.endStruct()
+              j += 1
+            }
+            rvb.endArray()
+
+            rvb.startArray(nCols)
+
+            val entriesOffset = localRVRowType.loadField(rv, entriesIndex)
+            val prevEntriesOffsets = rvs.map(localRVRowType.loadField(_, entriesIndex))
+
+            j = 0
+            while (j < nCols) {
+              rvb.startStruct()
+              if (entryArrayType.isElementDefined(rv.region, entriesOffset, j))
+                rvb.addAllFields(entryType, rv.region, entryArrayType.loadElement(rv.region, entriesOffset, j))
+              else
+                rvb.skipFields(entryType.size)
+
+              // prev_entries
+              rvb.startArray(rvs.length)
+              var k = 0
+              while (k < rvs.length) {
+                rvb.startStruct()
+                if (entryArrayType.isElementDefined(rvs(k).region, prevEntriesOffsets(k), j))
+                  rvb.addAllFields(entryType, rvs(k).region, entryArrayType.loadElement(rvs(k).region, prevEntriesOffsets(k), j))
+                else
+                  rvb.skipFields(entryType.size)
+                rvb.endStruct()
+                k += 1
+              }
+              rvb.endArray()
+              rvb.endStruct()
+
+              j += 1
+            }
+            rvb.endArray()
+            rvb.endStruct()
+
+            rv2.set(region, rvb.end())
+            Some(rv2)
+          }
+
+          if (!inAdjustment || rg.compare(locusStartsBc.value(partitionIndex), locus) <= 0) {
```

Comment:
can't you move this into the `else` above and just check the second condition? And then you also don't need `val result = if ...`, just the `if else`

Reply:
I need to do this for rows that are in adjustments in the window, though. and I want to do it at the bottom so I don't need to worry about dropping the current thing from the queue